src/execution/operator/join/physical_iejoin.cpp

```diff
@@ -0,0 +1,1415 @@
+
+			right_types.push_back(cond.left->return_type);
+			right_executor.AddExpression(*cond.right);
+		}
+
+		left_keys.Initialize(left_types);
+		right_keys.Initialize(right_types);
+	}
+
+	idx_t SelectJoinTail(const ExpressionType &condition, Vector &left, Vector &right, const SelectionVector *sel,
+	                     idx_t count);
+
+	idx_t SelectOuterRows(bool *matches) {
+		idx_t count = 0;
+		for (; outer_idx < outer_count; ++outer_idx) {
+			if (!matches[outer_idx]) {
+				true_sel.set_index(count++, outer_idx);
+				if (count >= STANDARD_VECTOR_SIZE) {
+					break;
+				}
+			}
+		}
+
+		return count;
+	}
+
+	const PhysicalIEJoin &op;
+
+	// Joining
+	unique_ptr<IEJoinUnion> joiner;
+
+	idx_t left_base;
+	idx_t left_block_index;
+
+	idx_t right_base;
+	idx_t right_block_index;
+
+	// Trailing predicates
+	SelectionVector true_sel;
+
+	ExpressionExecutor left_executor;
+	DataChunk left_keys;
+
+	ExpressionExecutor right_executor;
+	DataChunk right_keys;
+
+	// Outer joins
+	idx_t outer_idx;
+	idx_t outer_count;
+	bool *left_matches;
+	bool *right_matches;
+};
+
+idx_t IEJoinLocalSourceState::SelectJoinTail(const ExpressionType &condition, Vector &left, Vector &right,
+                                             const SelectionVector *sel, idx_t count) {
+	switch (condition) {
+	case ExpressionType::COMPARE_NOTEQUAL:
+		return VectorOperations::NotEquals(left, right, sel, count, &true_sel, nullptr);
+	case ExpressionType::COMPARE_LESSTHAN:
+		return VectorOperations::LessThan(left, right, sel, count, &true_sel, nullptr);
+	case ExpressionType::COMPARE_GREATERTHAN:
+		return VectorOperations::GreaterThan(left, right, sel, count, &true_sel, nullptr);
+	case ExpressionType::COMPARE_LESSTHANOREQUALTO:
+		return VectorOperations::LessThanEquals(left, right, sel, count, &true_sel, nullptr);
+	case ExpressionType::COMPARE_GREATERTHANOREQUALTO:
+		return VectorOperations::GreaterThanEquals(left, right, sel, count, &true_sel, nullptr);
+	case ExpressionType::COMPARE_DISTINCT_FROM:
+		return VectorOperations::DistinctFrom(left, right, sel, count, &true_sel, nullptr);
+	case ExpressionType::COMPARE_NOT_DISTINCT_FROM:
+	case ExpressionType::COMPARE_EQUAL:
+	default:
+		throw InternalException("Unsupported comparison type for PhysicalIEJoin");
+	}
+
+	return count;
+}
+
+void PhysicalIEJoin::ResolveComplexJoin(ExecutionContext &context, DataChunk &chunk, LocalSourceState &state_p) const {
+	auto &state = (IEJoinLocalSourceState &)state_p;
+	auto &ie_sink = (IEJoinGlobalState &)*sink_state;
+	auto &left_table = *ie_sink.tables[0];
+	auto &right_table = *ie_sink.tables[1];
+
+	const auto left_cols = children[0]->GetTypes().size();
+	do {
+		SelectionVector lsel(STANDARD_VECTOR_SIZE);
+		SelectionVector rsel(STANDARD_VECTOR_SIZE);
+		auto result_count = state.joiner->JoinComplexBlocks(lsel, rsel);
+		if (result_count == 0) {
+			// exhausted this pair
+			return;
+		} else {
+			// found matches: extract them
+			chunk.Reset();
```

Comment:
nit: this else can go to reduce indentation depth here

Reply:
Ah, there used to be a body in the return - sorry!